### DISCUSSION TITLE
Is it possible define tuples instead of lists for types?

### DISCUSSION THREAD

**Discussion by nbro10** [koxudaxi/datamodel-code-generator#2132] [General]
I have a json schema like this

```
{
    "type": "object",
    "description": "My desc",
    "properties": {
        "x": {
            "$ref": "./A.json#"
        },
        "y": {
            "type": "array",
            "minItems": 2,
            "maxItems": 2,
            "items": [
                {
                    "$ref": "./B.json#"
                },
                {
                    "oneOf": [
                        {
                            "$ref": "./C.json#"
                        },
                        {
                            "$ref": "./D.json#"
                        }
                    ]
                }
            ]
        }
    },
    "required": [
        "x",
        "y"
    ],
    "additionalProperties": false
}
```

This generates a Pydantic V2 model like this (or something like this)

```
class Model(BaseModel):
    """
    My Desc
    """

    model_config = ConfigDict(
        extra="forbid",
    )
    x: A
    y: list[B | C| D] = Field(
        ..., max_length=2, min_length=2
    )

```

But what I want is that `y` is a tuple, where the first element is of type `B` and the second element can be `C` or `D`. So, I want this 

```
class Model(BaseModel):
    """
    My Desc
    """

    model_config = ConfigDict(
        extra="forbid",
    )
    x: A
   y: tuple[B, C | D] = Field(
        ..., max_length=2, min_length=2
    )
```

In other words, the first element should always be a `B`, while the second can be a `C` or `D`. In reality, I don't really care if this is a tuple, it could also be a list, as long as the type constraints of the elements I've just described is satisfied.

I don't know if this is a JSON schema limitation or not.

I thought of overriding the template, but maybe there's an easier solution just by changing my schema to specify that?

Also, I don't really know right now how I'd need to override the template.

**Comment by nbro10** [comment]
So, my understanding is that this should be possible with `prefixItems`, but this library still doesn't support this feature: https://github.com/koxudaxi/datamodel-code-generator/issues/1546. In fact, if I manually change the list type to the tuple I want and use Pydantic to generate the schema, it generates a schema like mine except that it uses `prefixItems` instead of `items`.

**Reply by nbro10** [reply]
So, `prefixItems` was added only in JSON schema draft 2020. I don't know which versions of JSON schema `datamodel-code-generator` supports. There are other tools like `hypothesis-schema` that don't support 2020. It's a pity because these versions have been around for 4 years (I suppose).

However, according to https://github.com/orgs/json-schema-org/discussions/535#discussioncomment-7613705, I think we don't strictly need `prefixItems` to specify that each item in the array has a different type or subschema. 

So, I think that `datamodel-code-generator` should provide a flag that replaces `lists` with `tuples` at least in specific cases, like when the min and max number of elements in the array are equal and we're using different types or schemas for the elements. In that case, we most likely want to use tuples instead of lists.

**Comment by koxudaxi** [comment]
This feature is already supported — when using prefixItems with minItems and maxItems equal to the length of prefixItems, tuple types are automatically generated (e.g., tuple[Span, str]). Thank you!!

**Reply by nbro10** [reply]
Thanks for implementing this and all other features for which I had opened an issue. I really appreciate it. I've not yet tried them, because I'm still using an old version of the library, but I'll try the newest version asap and let you know.
